### DISCUSSION TITLE
add signature provider to OpenSSL

### DISCUSSION THREAD

**Discussion by ArtinKennedy** [openssl/openssl#23829] [Q&A] [answered]
Hi
I want to add a signature provider to the OpenSSL library. Since "discover_provider_groups" function return the group from capabilities, I get an error in ssl_load_groups function
(https://github.com/openssl/openssl/blob/5677992679b38950c6a0c3775fd57378e1879ba5/ssl/ssl_lib.c#L3923).
Is it not possible to add only signature algorithm in the provider?

**Comment by mattcaswell** [comment]
> Since "discover_provider_groups" function return the group from capabilities, I get an error in ssl_load_groups function

It's unclear to me what you are doing and what you expect to happen. Did you implement the "TLS-GROUP" capability in your provider or not? If you do not support any capabilities then I would not expect to see an error from `ssl_load_groups`. If you implemented the capability then I would only expect `ssl_load_groups` to fail if the capability implementation in your provider returned an error.

**Reply by ArtinKennedy** [reply]
i want to add only a signature provider like Dilithium signature and i do not want to add Kem (group) like kyber.
when i do that i got error.
this error occur because of this code:
```
static int discover_provider_groups(OSSL_PROVIDER *provider, void *vctx)
{
    struct provider_ctx_data_st pgd;

    pgd.ctx = vctx;
    pgd.provider = provider;
    return OSSL_PROVIDER_get_capabilities(provider, "TLS-GROUP",
                                          add_provider_groups, &pgd);
}
```
this code always return output of **OSSL_PROVIDER_get_capabilities** and in **my provider** i do not have **TLS-GROUP**.
i only have **TLS-SIGALG**.

**Reply by mattcaswell** [reply]
Ah, I see! Hmm. What does your provider return in the event of an unrecognised capability? I'm assuming it returns 0. Try returning 1 instead?

**Reply by ArtinKennedy** [reply]
at the first time i use this code and i got error:
```
int oqs_provider_get_capabilities(void *provctx, const char *capability,
                              OSSL_CALLBACK *cb, void *arg)
{
  if (strcasecmp(capability, "TLS-SIGALG") == 0)
        return oqs_sig_capability(cb, arg);

    /* don't support this capability */
    return 0;
}
```
and at the second time i use this code and it is work:

```
int oqs_provider_get_capabilities(void *provctx, const char *capability,
                              OSSL_CALLBACK *cb, void *arg)
{
  if (strcasecmp(capability, "TLS-GROUP") == 0)
        return 1;

  if (strcasecmp(capability, "TLS-SIGALG") == 0)
        return oqs_sig_capability(cb, arg);

    /* don't support this capability */
    return 0;
}
```

**Reply by ArtinKennedy** [reply]
i think this code is working too :
```
int oqs_provider_get_capabilities(void *provctx, const char *capability,
                              OSSL_CALLBACK *cb, void *arg)
{

  if (strcasecmp(capability, "TLS-SIGALG") == 0)
        return oqs_sig_capability(cb, arg);

    /* don't support this capability */
    return 1;
}
```

**Reply by ArtinKennedy** [reply]
**_**but my question is why this error happen ?**_**
this happen because of the default provider (group like secp521r1) ?

**Reply by mattcaswell** [reply, marked as answer]
It happens because "return 0" from the get_capabilities call is interpreted as a fatal error and OpenSSL gives up. By returning 1 we turn that into "we successfully processed the get_capabilities call, but we have nothing to do". I notice OpenSSL's own providers also return 0 for unrecognised capabilities which sounds like we might be storing up a problem for ourselves in the future when we add new capabilities. We should probably fix that.
